### PR TITLE
AEA-3604 Clinical Prescription Tracker API AWS

### PR DESCRIPTION
**Pull request by kris-szlapa**
# Description

:pencil: Summary:
This pull request proposes the creation of a new repository named "electronic-prescription-service-clinical-prescription-tracker" to establish the foundation for the **Clinical Prescription Tracker API**. The repository structure and initial content will mirror that of the existing "prescriptionsforpatients" repository. It will include a basic stack for CI and release builds of the Clinical Prescription Tracker API.

:ticket: Jira References: 
- [AEA-3604](https://nhsd-jira.digital.nhs.uk/browse/AEA-3604) create the Clinical Prescription Tracker API AWS repo
- [AEA-3677](https://nhsd-jira.digital.nhs.uk/browse/AEA-3677) deploy basic stack for CI and release builds of the Clinical Prescription Tracker API

### CODE REVIEW

File: .github/scripts/call_mark_jira_released.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
+
+echo "calling mark jira released"
+
+cat <<EOF > payload.json
+{ 
+  "releaseVersion": "PfP-AWS-$RELEASE_TAG"
+}
```
Comment (anthony-nhs) [inline .github/scripts/call_mark_jira_released.sh:8]: this should be  
```
  "releaseVersion": "Clinical-Tracker-$RELEASE_TAG"
```

File: .github/scripts/create_env_release_notes.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+cat <<EOF > payload.json
+{ 
+  "currentTag": "$CURRENT_DEPLOYED_TAG",
+  "targetTag": "$DEV_TAG",
+  "repoName": "prescriptionsforpatients",
+  "targetEnvironment": "$ENV",
+  "productName": "Prescriptions for Patients AWS layer",
+  "releaseNotesPageId": "$PAGE_ID",
+  "releaseNotesPageTitle": "Current PfP AWS layer release notes - $ENV"
+}
```
Comment (anthony-nhs) [inline .github/scripts/create_env_release_notes.sh:12]: this should be
```
{ 
  "currentTag": "$CURRENT_DEPLOYED_TAG",
  "targetTag": "$DEV_TAG",
  "repoName": "electronic-prescription-service-clinical-prescription-tracker",
  "targetEnvironment": "$ENV",
  "productName": "Clinical Tracker",
  "releaseNotesPageId": "$PAGE_ID",
  "releaseNotesPageTitle": "Current Clinical Tracker release notes - $ENV"
}
```

File: .github/scripts/create_int_rc_release_notes.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  "targetTag": "$RELEASE_TAG",
+  "repoName": "prescriptionsforpatients",
+  "targetEnvironment": "INT",
+  "productName": "Prescriptions for Patients AWS layer",
+  "releaseNotesPageId": "$PAGE_ID",
+  "releaseNotesPageTitle": "PfP-AWS-$RELEASE_TAG - Deployed to [INT] on $(date +'%d-%m-%y')",
+  "createReleaseCandidate": "true",
+  "releasePrefix": "PfP-AWS-"
+}
+EOF
```
Comment (anthony-nhs) [inline .github/scripts/create_int_rc_release_notes.sh:15]: this should be 
```
 
  "currentTag": "$CURRENT_DEPLOYED_TAG",
  "targetTag": "$RELEASE_TAG",
  "repoName": "electronic-prescription-service-clinical-prescription-tracker",
  "targetEnvironment": "INT",
  "productName": "Clinical Tracker",
  "releaseNotesPageId": "$PAGE_ID",
  "releaseNotesPageTitle": "Clinical-Tracker-$RELEASE_TAG - Deployed to [INT] on $(date +'%d-%m-%y')",
  "createReleaseCandidate": "true",
  "releasePrefix": "Clinical-Tracker-"
}
```

File: .github/scripts/delete_stacks.sh [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+
+ACTIVE_STACKS=$(aws cloudformation list-stacks | jq -r '.StackSummaries[] | select ( .StackStatus != "DELETE_COMPLETE" ) | select( .StackName | capture("^pr-(sandbox-)?(\\d+)$") ) | .StackName ')
+
```
Comment (anthony-nhs) [inline .github/scripts/delete_stacks.sh:4]: this should be 
```
ACTIVE_STACKS=$(aws cloudformation list-stacks | jq -r '.StackSummaries[] | select ( .StackStatus != "DELETE_COMPLETE" ) | select( .StackName | capture("^clinical-tracker-pr-(sandbox-)?(\\d+)$") ) | .StackName ')
```

File: .github/scripts/delete_stacks.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+mapfile -t ACTIVE_STACKS_ARRAY <<< "$ACTIVE_STACKS"
+
+for i in "${ACTIVE_STACKS_ARRAY[@]}"
+do 
+  echo "Checking if stack $i has open pull request"
+  PULL_REQUEST=${i//pr-/}
+  PULL_REQUEST=${PULL_REQUEST//sandbox-/}
+  echo "Checking pull request id ${PULL_REQUEST}"
+  URL="https://api.github.com/repos/NHSDigital/prescriptionsforpatients/pulls/${PULL_REQUEST}"
```
Comment (anthony-nhs) [inline .github/scripts/delete_stacks.sh:13]: this should be 
```
  URL="https://api.github.com/repos/NHSDigital/electronic-prescription-service-clinical-prescription-tracker/pulls/${PULL_REQUEST}"
```

File: .github/scripts/get_current_dev_tag.sh [hunk 6]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+dev_tag=$(aws cloudformation describe-stacks --stack-name dev-ci --query "Stacks[0].Tags[?Key=='version'].Value" --output text)
+
```
Comment (anthony-nhs) [inline .github/scripts/get_current_dev_tag.sh:4]: this should be 
```
dev_tag=$(aws cloudformation describe-stacks --stack-name clinical-tracker --query "Stacks[0].Tags[?Key=='version'].Value" --output text)
```

File: .github/scripts/get_target_deployed_tag.sh [hunk 7]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+current_deployed_tag=$(aws cloudformation describe-stacks --stack-name "$TARGET_ENVIRONMENT"-ci --query "Stacks[0].Tags[?Key=='version'].Value" --output text)
```
Comment (anthony-nhs) [inline .github/scripts/get_target_deployed_tag.sh:3]: this should be 
```
current_deployed_tag=$(aws cloudformation describe-stacks --stack-name clinical-tracker --query "Stacks[0].Tags[?Key=='version'].Value" --output text)

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+  package_code:
+    needs: tag_release
+    uses: ./.github/workflows/sam_package_code.yml
+
+  release_dev:
+    needs: [tag_release, package_code, get_commit_id]
+    uses: ./.github/workflows/sam_release_code.yml
+    with:
+      ARTIFACT_BUCKET_PREFIX: ${{needs.tag_release.outputs.spec_version}}
+      STACK_NAME: dev-ci
```
Comment (anthony-nhs) [inline .github/workflows/release.yml:107]: this should be 
```
      STACK_NAME: clinical-tracker
```

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+      INT_CLOUD_FORMATION_CHECK_VERSION_ROLE: ${{ secrets.INT_CLOUD_FORMATION_CHECK_VERSION_ROLE }}
+      PROD_CLOUD_FORMATION_CHECK_VERSION_ROLE: ${{ secrets.PROD_CLOUD_FORMATION_CHECK_VERSION_ROLE }}
+      DEV_CLOUD_FORMATION_EXECUTE_LAMBDA_ROLE: ${{ secrets.DEV_CLOUD_FORMATION_EXECUTE_LAMBDA_ROLE }}
+
+  release_dev_sandbox:
+    needs: [tag_release, package_code, get_commit_id]
+    uses: ./.github/workflows/sam_release_code.yml
+    with:
+      ARTIFACT_BUCKET_PREFIX: ${{needs.tag_release.outputs.spec_version}}
+      STACK_NAME: dev-sandbox
```
Comment (anthony-nhs) [inline .github/workflows/release.yml:132]: this should be 
```
      STACK_NAME: sandbox-clinical-tracker
```

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    secrets:
+      CLOUD_FORMATION_DEPLOY_ROLE: ${{ secrets.DEV_CLOUD_FORMATION_DEPLOY_ROLE }}
+      TARGET_SPINE_SERVER: sandbox
+
+  release_ref:
+    needs: [tag_release, release_dev, release_dev_sandbox, package_code, get_commit_id]
+    uses: ./.github/workflows/sam_release_code.yml
+    with:
+      ARTIFACT_BUCKET_PREFIX: ${{needs.tag_release.outputs.spec_version}}
+      STACK_NAME: ref-ci
```
Comment (anthony-nhs) [inline .github/workflows/release.yml:151]: this should be 
```
      STACK_NAME: clinical-tracker
```

File: .github/workflows/release.yml [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+    secrets:
+      CLOUD_FORMATION_DEPLOY_ROLE: ${{ secrets.REF_CLOUD_FORMATION_DEPLOY_ROLE }}
+      TARGET_SPINE_SERVER: ${{ secrets.REF_TARGET_SPINE_SERVER }}
+
+  release_qa:
+    needs: [tag_release, release_dev, release_dev_sandbox, package_code, get_commit_id]
+    uses: ./.github/workflows/sam_release_code.yml
+    with:
+      ARTIFACT_BUCKET_PREFIX: ${{needs.tag_release.outputs.spec_version}}
+      STACK_NAME: qa-ci
```
Comment (anthony-nhs) [inline .github/workflows/release.yml:170]: this should be 
```
      STACK_NAME: clinical-tracker
```

File: .github/workflows/release.yml [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+    secrets:
+      CLOUD_FORMATION_DEPLOY_ROLE: ${{ secrets.QA_CLOUD_FORMATION_DEPLOY_ROLE }}
+      TARGET_SPINE_SERVER: ${{ secrets.QA_TARGET_SPINE_SERVER }}
+
+  release_int:
+    needs: [tag_release, release_qa, package_code, get_commit_id]
+    uses: ./.github/workflows/sam_release_code.yml
+    with:
+      ARTIFACT_BUCKET_PREFIX: ${{needs.tag_release.outputs.spec_version}}
+      STACK_NAME: int-ci
```
Comment (anthony-nhs) [inline .github/workflows/release.yml:189]: this should be 
```
      STACK_NAME: clinical-tracker
```

File: .github/workflows/release.yml [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+      INT_CLOUD_FORMATION_CHECK_VERSION_ROLE: ${{ secrets.INT_CLOUD_FORMATION_CHECK_VERSION_ROLE }}
+      PROD_CLOUD_FORMATION_CHECK_VERSION_ROLE: ${{ secrets.PROD_CLOUD_FORMATION_CHECK_VERSION_ROLE }}
+      DEV_CLOUD_FORMATION_EXECUTE_LAMBDA_ROLE: ${{ secrets.DEV_CLOUD_FORMATION_EXECUTE_LAMBDA_ROLE }}
+
+  release_sandbox:
+    needs: [tag_release, release_qa, package_code, get_commit_id]
+    uses: ./.github/workflows/sam_release_code.yml
+    with:
+      ARTIFACT_BUCKET_PREFIX: ${{needs.tag_release.outputs.spec_version}}
+      STACK_NAME: int-sandox
```
Comment (anthony-nhs) [inline .github/workflows/release.yml:214]: this should be 
```
      STACK_NAME: sandbox-clinical-tracker
```

File: .github/workflows/release.yml [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+    secrets:
+      CLOUD_FORMATION_DEPLOY_ROLE: ${{ secrets.INT_CLOUD_FORMATION_DEPLOY_ROLE }}
+      TARGET_SPINE_SERVER: sandbox
+
+  release_prod:
+    needs: [tag_release, release_int, release_sandbox, package_code, get_commit_id]
+    uses: ./.github/workflows/sam_release_code.yml
+    with:
+      ARTIFACT_BUCKET_PREFIX: ${{needs.tag_release.outputs.spec_version}}
+      STACK_NAME: prod-ci
```
Comment (anthony-nhs) [inline .github/workflows/release.yml:233]: this should be 
```
      STACK_NAME: clinical-tracker
```

File: .github/workflows/sam_release_code.yml [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+          sparse-checkout: |
+            .github
+
+      - name: create_int_rc_release_notes
+        uses: ./.github/actions/update_confluence_jira
+        if: ${{ inputs.CREATE_INT_RC_RELEASE_NOTES == true }}
+        with:
+          TARGET_ENVIRONMENT: int
+          RELEASE_TAG: ${{ inputs.VERSION_NUMBER }}
+          CONFLUENCE_PAGE_ID: "710051481"
```
Comment (anthony-nhs) [inline .github/workflows/sam_release_code.yml:88]: this should be
```
          CONFLUENCE_PAGE_ID: "768067994"
```
its page https://nhsd-confluence.digital.nhs.uk/display/APIMC/Clinical+Tracker+Release+Cuts

File: .github/workflows/sam_release_code.yml [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+          VERSION_NUMBER: ${{ inputs.VERSION_NUMBER }}
+        run: |
+          ./release_code.sh
+
+      - name: create_int_release_notes
+        uses: ./.github/actions/update_confluence_jira
+        if: ${{ inputs.CREATE_INT_RELEASE_NOTES == true && always() && !failure() && !cancelled() }}
+        with:
+          TARGET_ENVIRONMENT: int
+          CONFLUENCE_PAGE_ID: "693750027"
```
Comment (anthony-nhs) [inline .github/workflows/sam_release_code.yml:131]: this should be 
```
          CONFLUENCE_PAGE_ID: "768067990"
```
its page https://nhsd-confluence.digital.nhs.uk/display/APIMC/Current+Clinical+Tracker+release+notes+-+INT

File: .github/workflows/sam_release_code.yml [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+          DEV_CLOUD_FORMATION_CHECK_VERSION_ROLE: ${{ secrets.DEV_CLOUD_FORMATION_CHECK_VERSION_ROLE }}
+          TARGET_CLOUD_FORMATION_CHECK_VERSION_ROLE: ${{ secrets.INT_CLOUD_FORMATION_CHECK_VERSION_ROLE }}
+          DEV_CLOUD_FORMATION_EXECUTE_LAMBDA_ROLE: ${{ secrets.DEV_CLOUD_FORMATION_EXECUTE_LAMBDA_ROLE }}
+
+      - name: create_prod_release_notes
+        uses: ./.github/actions/update_confluence_jira
+        if: ${{ inputs.CREATE_PROD_RELEASE_NOTES == true && always() && !failure() && !cancelled() }}
+        with:
+          TARGET_ENVIRONMENT: prod
+          CONFLUENCE_PAGE_ID: "693750029"
```
Comment (anthony-nhs) [inline .github/workflows/sam_release_code.yml:142]: this should be 
```
         CONFLUENCE_PAGE_ID: "768067992"
```
its page https://nhsd-confluence.digital.nhs.uk/display/APIMC/Current+Clinical+Tracker+release+notes+-+PROD

File: .github/scripts/create_env_release_notes.sh (hunk 2 above)
Comment (kris-szlapa) [reply .github/scripts/create_env_release_notes.sh:12]: The script has been updated

File: .github/scripts/create_int_rc_release_notes.sh (hunk 3 above)
Comment (kris-szlapa) [reply .github/scripts/create_int_rc_release_notes.sh:15]: The script has been updated

File: .github/scripts/delete_stacks.sh (hunk 4 above)
Comment (kris-szlapa) [reply .github/scripts/delete_stacks.sh:4]: The script has been updated

File: .github/scripts/delete_stacks.sh (hunk 5 above)
Comment (kris-szlapa) [reply .github/scripts/delete_stacks.sh]: The URL has been updated

File: .github/scripts/get_target_deployed_tag.sh (hunk 7 above)
Comment (kris-szlapa) [reply .github/scripts/get_target_deployed_tag.sh]: The script has been updated

File: .github/scripts/get_current_dev_tag.sh (hunk 6 above)
Comment (kris-szlapa) [reply .github/scripts/get_current_dev_tag.sh:4]: The script has been updated

File: .github/workflows/release.yml (hunk 8 above)
Comment (kris-szlapa) [reply .github/workflows/release.yml]: The STACK_NAME value has been changed

File: .github/workflows/release.yml (hunk 9 above)
Comment (kris-szlapa) [reply .github/workflows/release.yml]: The STACK_NAME value has been changed

File: .github/workflows/release.yml (hunk 10 above)
Comment (kris-szlapa) [reply .github/workflows/release.yml]: The STACK_NAME value has been changed

File: .github/workflows/release.yml (hunk 11 above)
Comment (kris-szlapa) [reply .github/workflows/release.yml]: The STACK_NAME value has been changed

File: .github/workflows/release.yml (hunk 12 above)
Comment (kris-szlapa) [reply .github/workflows/release.yml]: The STACK_NAME value has been changed

File: .github/workflows/release.yml (hunk 13 above)
Comment (kris-szlapa) [reply .github/workflows/release.yml]: The STACK_NAME value has been changed

File: .github/workflows/release.yml (hunk 14 above)
Comment (kris-szlapa) [reply .github/workflows/release.yml]: The STACK_NAME value has been changed

File: .github/workflows/sam_release_code.yml (hunk 16 above)
Comment (kris-szlapa) [reply .github/workflows/sam_release_code.yml]: The CONFLUENCE_PAGE_ID value has been updated

File: .github/workflows/sam_release_code.yml (hunk 17 above)
Comment (kris-szlapa) [reply .github/workflows/sam_release_code.yml]: The CONFLUENCE_PAGE_ID value has been updated

**Pull request merged**